File: Sources/M3U8DownloaderKitObjC/M3U8Downloader.m
//
//  M3U8Downloader.m
//  M3U8DownloaderKit
//
//  主下载器类实现
//

#import "M3U8Downloader.h"

// 直接导入 C 头文件（ObjC 与 C 完全兼容）
#include "m3u8dl.h"

#pragma mark - M3U8Downloader 私有接口（前向声明）

// 必须在 C 回调函数之前声明，否则编译器找不到方法
@interface M3U8Downloader ()

/// 内部实例 ID
@property (nonatomic, assign) m3u8dl_instance_t instanceId;

/// 是否已释放
@property (nonatomic, assign, readwrite) BOOL isDisposed;

/// 当前状态
@property (nonatomic, assign, readwrite) M3U8DownloadStatus currentStatus;

/// 进度回调 Block
@property (nonatomic, copy, nullable) M3U8ProgressBlock progressBlock;

/// 完成回调 Block
@property (nonatomic, copy, nullable) M3U8DownloadCompletionBlock completionBlock;

/// 日志回调 Block
@property (nonatomic, copy, nullable) M3U8LogBlock logBlock;

/// 状态锁
@property (nonatomic, strong) NSLock *lock;

/// 处理进度回调
- (void)handleProgress:(M3U8DownloadProgress *)progress;

/// 处理完成回调
- (void)handleCompletion:(M3U8DownloadResult *)result;

/// 处理日志回调
- (void)handleLog:(M3U8LogLevel)level message:(NSString *)message;

@end

#pragma mark - 静态变量

/// 全局实例映射
static NSMutableDictionary<NSNumber *, M3U8Downloader *> *s_instances = nil;

/// 实例映射锁
static NSLock *s_instancesLock = nil;

/// 回调是否已设置
static BOOL s_callbacksSetup = NO;

#pragma mark - C 回调函数

static void globalProgressCallback(m3u8dl_instance_t instanceId, int32_t percentage, const char *statusJson) {
    if (!statusJson) return;
    
    NSString *jsonString = [NSString stringWithUTF8String:statusJson];
    M3U8DownloadProgress *progress = [M3U8DownloadProgress parseFromJSONString:jsonString];
    if (!progress) return;
    
    // 更新百分比（C API 传递的）
    progress.percentage = percentage;
    
    [s_instancesLock lock];
    M3U8Downloader *instance = s_instances[@(instanceId)];
    [s_instancesLock unlock];
    
    if (instance) {
        [instance handleProgress:progress];
    }
}

static void globalCompletionCallback(m3u8dl_instance_t instanceId, m3u8dl_bool_t success, const char *resultJson) {
    if (!resultJson) return;
    
    NSString *jsonString = [NSString stringWithUTF8String:resultJson];
    M3U8DownloadResult *result = [M3U8DownloadResult parseFromJSONString:jsonString];
    if (!result) {
        result = [[M3U8DownloadResult alloc] init];
        result.success = (success != 0);
    }
    
    [s_instancesLock lock];
    M3U8Downloader *instance = s_instances[@(instanceId)];
    [s_instancesLock unlock];
    
    if (instance) {
        [instance handleCompletion:result];
    }
}

static void globalLogCallback(m3u8dl_log_level_t level, const char *message) {
    if (!message) return;
    
    NSString *messageString = [NSString stringWithUTF8String:message];
    M3U8LogLevel logLevel = (M3U8LogLevel)level;
    
    // 分发到所有实例
    [s_instancesLock lock];
    NSArray *allInstances = [s_instances.allValues copy];
    [s_instancesLock unlock];
    
    for (M3U8Downloader *instance in allInstances) {
        [instance handleLog:logLevel message:messageString];
    }
}

#pragma mark - M3U8Downloader 实现

@implementation M3U8Downloader

#pragma mark - 类方法

+ (void)initialize {
    if (self == [M3U8Downloader class]) {
        s_instances = [NSMutableDictionary dictionary];
        s_instancesLock = [[NSLock alloc] init];
    }
}

+ (void)setupGlobalCallbacks {
    if (s_callbacksSetup) return;
    s_callbacksSetup = YES;
    
    m3u8dl_set_progress_callback(globalProgressCallback);
    m3u8dl_set_completion_callback(globalCompletionCallback);
    m3u8dl_set_log_callback(globalLogCallback);
}

+ (void)registerInstance:(M3U8Downloader *)instance {
    [s_instancesLock lock];
    s_instances[@(instance.instanceId)] = instance;
    [s_instancesLock unlock];
}

+ (void)unregisterInstance:(m3u8dl_instance_t)instanceId {
    [s_instancesLock lock];
    [s_instances removeObjectForKey:@(instanceId)];
    [s_instancesLock unlock];
}

+ (M3U8VersionInfo *)versionInfo {
    char *resultPtr = m3u8dl_get_version();
    if (!resultPtr) return nil;
    
    NSString *jsonString = [NSString stringWithUTF8String:resultPtr];
    m3u8dl_free_string(resultPtr);
    
    return [M3U8VersionInfo parseFromJSONString:jsonString];
}

+ (void)disposeAll {
    m3u8dl_dispose_all();
    
    [s_instancesLock lock];
    for (M3U8Downloader *instance in s_instances.allValues) {
        instance.isDisposed = YES;
    }
    [s_instances removeAllObjects];
    [s_instancesLock unlock];
}

#pragma mark - 初始化

- (instancetype)initWithError:(NSError **)error {
    return [self initWithConfiguration:nil error:error];
}

- (instancetype)initWithConfiguration:(M3U8Configuration *)configuration error:(NSError **)error {
    self = [super init];
    if (self) {
        // 设置全局回调
        [[self class] setupGlobalCallbacks];
        
        _lock = [[NSLock alloc] init];
        _isDisposed = NO;
        _currentStatus = M3U8DownloadStatusIdle;
        
        // 初始化 C 库
        const char *configJson = NULL;
        NSString *configString = [configuration toJSONString];
        if (configString) {
            configJson = [configString UTF8String];
        }
        
        m3u8dl_instance_t instanceId = m3u8dl_init(configJson);
        if (instanceId <= 0) {
            if (error) {
                *error = M3U8ErrorMake(M3U8ErrorCodeInitializationFailed, @"Failed to initialize C library");
            }
            return nil;
        }
        
        _instanceId = instanceId;
        
        // 注册实例
        [[self class] registerInstance:self];
    }
    return self;
}

- (void)dealloc {
    [self dispose];
}

#pragma mark - 解析方法

- (M3U8ParseResult *)parseURL:(NSString *)url error:(NSError **)error {
    if (![self checkDisposed:error]) return nil;
    
    if (!url || url.length == 0) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeInvalidParameter, @"URL cannot be empty");
        }
        return nil;
    }
    
    char *resultPtr = m3u8dl_parse(self.instanceId, [url UTF8String]);
    if (!resultPtr) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeParseFailed, @"Parse returned no result");
        }
        return nil;
    }
    
    NSString *jsonString = [NSString stringWithUTF8String:resultPtr];
    m3u8dl_free_string(resultPtr);
    
    M3U8ParseResult *result = [M3U8ParseResult parseFromJSONString:jsonString];
    if (!result) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeJSONError, @"Failed to parse JSON response");
        }
        return nil;
    }
    
    if (!result.success) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeParseFailed, result.errorMessage ?: @"Unknown error");
        }
        return nil;
    }
    
    return result;
}

- (void)parseURL:(NSString *)url completion:(M3U8ParseCompletionBlock)completion {
    if (!completion) return;
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSError *error = nil;
        M3U8ParseResult *result = [self parseURL:url error:&error];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completion(result, error);
        });
    });
}

#pragma mark - 下载方法

- (M3U8DownloadResult *)downloadURL:(NSString *)url
                             toPath:(NSString *)outputPath
                            options:(M3U8DownloadOptions *)options
                              error:(NSError **)error {
    if (![self checkDisposed:error]) return nil;
    
    if (!url || url.length == 0) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeInvalidParameter, @"URL cannot be empty");
        }
        return nil;
    }
    
    if (!outputPath || outputPath.length == 0) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeInvalidParameter, @"Output path cannot be empty");
        }
        return nil;
    }
    
    // 构建请求 JSON
    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"url"] = url;
    request[@"outputPath"] = outputPath;
    
    if (options) {
        request[@"autoSelectBestQuality"] = @(options.autoSelectBestQuality);
        if (options.customHeaders) {
            request[@"customHeaders"] = options.customHeaders;
        }
        if (options.decryptionKeys) {
            request[@"decryptionKeys"] = options.decryptionKeys;
        }
    }
    
    NSError *jsonError = nil;
    NSData *requestData = [NSJSONSerialization dataWithJSONObject:request options:0 error:&jsonError];
    if (jsonError || !requestData) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeJSONError, @"Failed to serialize request");
        }
        return nil;
    }
    
    NSString *requestJson = [[NSString alloc] initWithData:requestData encoding:NSUTF8StringEncoding];
    
    [self.lock lock];
    self.currentStatus = M3U8DownloadStatusDownloading;
    [self.lock unlock];
    
    char *resultPtr = m3u8dl_download(self.instanceId, [requestJson UTF8String]);
    
    [self.lock lock];
    self.currentStatus = M3U8DownloadStatusIdle;
    [self.lock unlock];
    
    if (!resultPtr) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeDownloadFailed, @"Download returned no result");
        }
        return nil;
    }
    
    NSString *jsonString = [NSString stringWithUTF8String:resultPtr];
    m3u8dl_free_string(resultPtr);
    
    M3U8DownloadResult *result = [M3U8DownloadResult parseFromJSONString:jsonString];
    if (!result) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeJSONError, @"Failed to parse JSON response");
        }
        return nil;
    }
    
    if (!result.success) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeDownloadFailed, result.errorMessage ?: @"Unknown error");
        }
        return nil;
    }
    
    return result;
}

- (void)downloadURL:(NSString *)url
             toPath:(NSString *)outputPath
            options:(M3U8DownloadOptions *)options
           progress:(M3U8ProgressBlock)progressBlock
         completion:(M3U8DownloadCompletionBlock)completion {
    if (!completion) return;
    
    // 保存回调
    [self.lock lock];
    self.progressBlock = progressBlock;
    self.completionBlock = completion;
    [self.lock unlock];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSError *error = nil;
        M3U8DownloadResult *result = [self downloadURL:url toPath:outputPath options:options error:&error];
        
        [self.lock lock];
        self.progressBlock = nil;
        M3U8DownloadCompletionBlock savedCompletion = self.completionBlock;
        self.completionBlock = nil;
        [self.lock unlock];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            if (savedCompletion) {
                savedCompletion(result, error);
            }
        });
    });
}

#pragma mark - 控制方法

- (void)cancel {
    [self.lock lock];
    BOOL disposed = self.isDisposed;
    [self.lock unlock];
    
    if (disposed) return;
    
    m3u8dl_cancel(self.instanceId);
    
    [self.lock lock];
    self.currentStatus = M3U8DownloadStatusCancelled;
    [self.lock unlock];
}

- (void)dispose {
    [self.lock lock];
    if (self.isDisposed) {
        [self.lock unlock];
        return;
    }
    self.isDisposed = YES;
    m3u8dl_instance_t instanceId = self.instanceId;
    [self.lock unlock];
    
    [[self class] unregisterInstance:instanceId];
    m3u8dl_dispose(instanceId);
}

#pragma mark - 回调设置

- (void)setLogHandler:(M3U8LogBlock)logBlock {
    [self.lock lock];
    self.logBlock = logBlock;
    [self.lock unlock];
}

#pragma mark - 功能查询

- (BOOL)isFeatureSupported:(NSString *)feature {
    [self.lock lock];
    BOOL disposed = self.isDisposed;
    [self.lock unlock];
    
    if (disposed || !feature) return NO;
    
    return m3u8dl_is_feature_supported(self.instanceId, [feature UTF8String]) != 0;
}

- (NSArray<NSString *> *)getProcessors:(NSError **)error {
    if (![self checkDisposed:error]) return nil;
    
    char *resultPtr = m3u8dl_get_processors(self.instanceId);
    if (!resultPtr) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeInternalError, @"Failed to get processors");
        }
        return nil;
    }
    
    NSString *jsonString = [NSString stringWithUTF8String:resultPtr];
    m3u8dl_free_string(resultPtr);
    
    NSData *data = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeJSONError, @"Invalid JSON encoding");
        }
        return nil;
    }
    
    NSError *jsonError = nil;
    NSArray *processors = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
    if (jsonError || ![processors isKindOfClass:[NSArray class]]) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeJSONError, @"Failed to parse processors JSON");
        }
        return nil;
    }
    
    return processors;
}

#pragma mark - 私有方法

- (BOOL)checkDisposed:(NSError **)error {
    [self.lock lock];
    BOOL disposed = self.isDisposed;
    [self.lock unlock];
    
    if (disposed) {
        if (error) {
            *error = M3U8ErrorMake(M3U8ErrorCodeDisposed, @"Downloader has been disposed");
        }
        return NO;
    }
    return YES;
}

- (void)handleProgress:(M3U8DownloadProgress *)progress {
    [self.lock lock];
    self.currentStatus = progress.status;
    M3U8ProgressBlock block = self.progressBlock;
    id<M3U8DownloaderDelegate> delegate = self.delegate;
    [self.lock unlock];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (block) {
            block(progress);
        }
        if ([delegate respondsToSelector:@selector(downloader:didUpdateProgress:)]) {
            [delegate downloader:self didUpdateProgress:progress];
        }
    });
}

- (void)handleCompletion:(M3U8DownloadResult *)result {
    [self.lock lock];
    self.currentStatus = result.success ? M3U8DownloadStatusCompleted : M3U8DownloadStatusFailed;
    id<M3U8DownloaderDelegate> delegate = self.delegate;
    [self.lock unlock];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([delegate respondsToSelector:@selector(downloader:didFinishWithResult:)]) {
            [delegate downloader:self didFinishWithResult:result];
        }
    });
}

- (void)handleLog:(M3U8LogLevel)level message:(NSString *)message {
    [self.lock lock];
    M3U8LogBlock block = self.logBlock;
    id<M3U8DownloaderDelegate> delegate = self.delegate;
    [self.lock unlock];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (block) {
            block(level, message);
        }
        if ([delegate respondsToSelector:@selector(downloader:didLogWithLevel:message:)]) {
            [delegate downloader:self didLogWithLevel:level message:message];
        }
    });
}

@end

File: Sources/M3U8DownloaderKit/M3U8Downloader.swift
//
//  M3U8Downloader.swift
//  M3U8DownloaderKit
//
//  高级 Swift 包装层，封装 C API 接口
//  符合 Swift 6 并发模型，使用 async/await
//

import Foundation
import M3U8DownloaderKit_C

// MARK: - M3U8Downloader

/// M3U8/HLS/DASH 流媒体下载器
///
/// 这是一个线程安全的下载器类，提供异步 API 用于解析和下载流媒体内容。
///
/// ## 基本用法
///
/// ```swift
/// let downloader = M3U8Downloader()
///
/// // 解析流信息
/// let streams = try await downloader.parse(url: "https://example.com/playlist.m3u8")
///
/// // 下载最高质量
/// let result = try await downloader.download(
///     url: "https://example.com/playlist.m3u8",
///     to: "/path/to/output.mp4"
/// )
/// ```
///
/// ## 进度监控
///
/// ```swift
/// for await progress in downloader.progressStream {
///     print("下载进度: \(progress.percentage)%")
/// }
/// ```
public final class M3U8Downloader: @unchecked Sendable {
    
    // MARK: - Properties
    
    /// 内部实例 ID
    private var instanceId: m3u8dl_instance_t
    
    /// 是否已释放
    private var isDisposed = false
    
    /// 锁保护状态
    private let lock = NSLock()
    
    /// 进度回调存储
    private var progressHandler: ((DownloadProgress) -> Void)?
    
    /// 完成回调存储
    private var completionHandler: ((DownloadResult) -> Void)?
    
    /// 日志回调存储
    private var logHandler: ((LogLevel, String) -> Void)?
    
    /// 进度流的 continuation
    private var progressContinuation: AsyncStream<DownloadProgress>.Continuation?
    
    // MARK: - Static Properties
    
    /// 共享的回调管理器
    /// 使用 nonisolated(unsafe) 因为此属性通过外部锁机制保护
    nonisolated(unsafe) private static var callbackManager = CallbackManager()
    
    /// 全局实例映射 (线程安全，通过 instancesLock 保护)
    /// 使用 nonisolated(unsafe) 因为访问受 instancesLock 保护
    nonisolated(unsafe) private static var instances: [m3u8dl_instance_t: M3U8Downloader] = [:]
    private static let instancesLock = NSLock()
    
    // MARK: - Initialization
    
    /// 创建下载器实例
    ///
    /// - Parameter configuration: 下载器配置，传入 nil 使用默认配置
    /// - Throws: `M3U8Error.initializationFailed` 如果初始化失败
    public init(configuration: Configuration? = nil) throws {
        // 设置全局回调（仅首次）
        Self.setupGlobalCallbacks()
        
        let configJson: String?
        if let config = configuration {
            let encoder = JSONEncoder()
            configJson = String(data: try encoder.encode(config), encoding: .utf8)
        } else {
            configJson = nil
        }
        
        let id = m3u8dl_init(configJson)
        guard id > 0 else {
            throw M3U8Error.initializationFailed
        }
        
        self.instanceId = id
        
        // 注册实例
        Self.registerInstance(self)
    }
    
    deinit {
        dispose()
    }
    
    // MARK: - Instance Management
    
    private static func registerInstance(_ instance: M3U8Downloader) {
        instancesLock.lock()
        defer { instancesLock.unlock() }
        instances[instance.instanceId] = instance
    }
    
    private static func unregisterInstance(_ instanceId: m3u8dl_instance_t) {
        instancesLock.lock()
        defer { instancesLock.unlock() }
        instances.removeValue(forKey: instanceId)
    }
    
    private static func getInstance(for id: m3u8dl_instance_t) -> M3U8Downloader? {
        instancesLock.lock()
        defer { instancesLock.unlock() }
        return instances[id]
    }
    
    // MARK: - Global Callbacks Setup
    
    /// 回调是否已设置（通过 callbacksSetupLock 保护）
    nonisolated(unsafe) private static var callbacksSetup = false
    private static let callbacksSetupLock = NSLock()
    
    private static func setupGlobalCallbacks() {
        callbacksSetupLock.lock()
        defer { callbacksSetupLock.unlock() }
        
        guard !callbacksSetup else { return }
        callbacksSetup = true
        
        // 设置进度回调
        m3u8dl_set_progress_callback { instanceId, percentage, statusJson in
            guard let statusJson = statusJson else { return }
            let jsonString = String(cString: statusJson)
            
            if let instance = M3U8Downloader.getInstance(for: instanceId) {
                instance.handleProgress(percentage: Int(percentage), statusJson: jsonString)
            }
        }
        
        // 设置完成回调
        m3u8dl_set_completion_callback { instanceId, success, resultJson in
            guard let resultJson = resultJson else { return }
            let jsonString = String(cString: resultJson)
            
            if let instance = M3U8Downloader.getInstance(for: instanceId) {
                instance.handleCompletion(success: success != 0, resultJson: jsonString)
            }
        }
        
        // 设置日志回调
        m3u8dl_set_log_callback { level, message in
            guard let message = message else { return }
            let logLevel = LogLevel(rawValue: Int(level.rawValue)) ?? .info
            let messageString = String(cString: message)
            
            // 分发到所有实例
            M3U8Downloader.instancesLock.lock()
            let allInstances = Array(M3U8Downloader.instances.values)
            M3U8Downloader.instancesLock.unlock()
            
            for instance in allInstances {
                instance.handleLog(level: logLevel, message: messageString)
            }
        }
    }
    
    // MARK: - Callback Handlers
    
    private func handleProgress(percentage: Int, statusJson: String) {
        guard let data = statusJson.data(using: .utf8),
              let progress = try? JSONDecoder().decode(DownloadProgress.self, from: data) else {
            return
        }
        
        progressHandler?(progress)
        progressContinuation?.yield(progress)
    }
    
    private func handleCompletion(success: Bool, resultJson: String) {
        guard let data = resultJson.data(using: .utf8),
              let result = try? JSONDecoder().decode(DownloadResult.self, from: data) else {
            return
        }
        
        completionHandler?(result)
    }
    
    private func handleLog(level: LogLevel, message: String) {
        logHandler?(level, message)
    }
    
    // MARK: - Public API
    
    /// 解析 M3U8/DASH URL，获取可用流信息
    ///
    /// - Parameter url: 流媒体 URL
    /// - Returns: 解析结果，包含所有可用流信息
    /// - Throws: `M3U8Error` 如果解析失败
    ///
    /// ## 示例
    ///
    /// ```swift
    /// let result = try await downloader.parse(url: "https://example.com/playlist.m3u8")
    /// for stream in result.streams {
    ///     print("流: \(stream.type) - \(stream.resolution ?? "N/A")")
    /// }
    /// ```
    public func parse(url: String) async throws -> ParseResult {
        try checkDisposed()
        
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [instanceId] in
                let resultPtr = m3u8dl_parse(instanceId, url)
                defer {
                    if let ptr = resultPtr {
                        m3u8dl_free_string(ptr)
                    }
                }
                
                guard let ptr = resultPtr else {
                    continuation.resume(throwing: M3U8Error.parseFailed(reason: "No result returned"))
                    return
                }
                
                let jsonString = String(cString: ptr)
                guard let data = jsonString.data(using: .utf8) else {
                    continuation.resume(throwing: M3U8Error.parseFailed(reason: "Invalid JSON encoding"))
                    return
                }
                
                do {
                    let result = try JSONDecoder().decode(ParseResult.self, from: data)
                    if result.success {
                        continuation.resume(returning: result)
                    } else {
                        continuation.resume(throwing: M3U8Error.parseFailed(reason: result.errorMessage ?? "Unknown error"))
                    }
                } catch {
                    continuation.resume(throwing: M3U8Error.parseFailed(reason: error.localizedDescription))
                }
            }
        }
    }
    
    /// 下载流媒体内容
    ///
    /// - Parameters:
    ///   - url: 流媒体 URL
    ///   - outputPath: 输出文件路径
    ///   - options: 下载选项
    /// - Returns: 下载结果
    /// - Throws: `M3U8Error` 如果下载失败
    ///
    /// ## 示例
    ///
    /// ```swift
    /// let result = try await downloader.download(
    ///     url: "https://example.com/playlist.m3u8",
    ///     to: "/path/to/output.mp4",
    ///     options: .init(autoSelectBestQuality: true)
    /// )
    /// print("下载完成: \(result.outputFile ?? "N/A")")
    /// ```
    public func download(
        url: String,
        to outputPath: String,
        options: DownloadOptions = .default
    ) async throws -> DownloadResult {
        try checkDisposed()
        
        let request = DownloadRequest(
            url: url,
            outputPath: outputPath,
            decryptionKeys: options.decryptionKeys,
            customHeaders: options.customHeaders,
            autoSelectBestQuality: options.autoSelectBestQuality
        )
        
        let encoder = JSONEncoder()
        let requestData = try encoder.encode(request)
        guard let requestJson = String(data: requestData, encoding: .utf8) else {
            throw M3U8Error.invalidParameter(name: "request")
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async { [instanceId] in
                let resultPtr = m3u8dl_download(instanceId, requestJson)
                defer {
                    if let ptr = resultPtr {
                        m3u8dl_free_string(ptr)
                    }
                }
                
                guard let ptr = resultPtr else {
                    continuation.resume(throwing: M3U8Error.downloadFailed(reason: "No result returned"))
                    return
                }
                
                let jsonString = String(cString: ptr)
                guard let data = jsonString.data(using: .utf8) else {
                    continuation.resume(throwing: M3U8Error.downloadFailed(reason: "Invalid JSON encoding"))
                    return
                }
                
                do {
                    let result = try JSONDecoder().decode(DownloadResult.self, from: data)
                    if result.success {
                        continuation.resume(returning: result)
                    } else {
                        continuation.resume(throwing: M3U8Error.downloadFailed(reason: result.errorMessage ?? "Unknown error"))
                    }
                } catch {
                    continuation.resume(throwing: M3U8Error.downloadFailed(reason: error.localizedDescription))
                }
            }
        }
    }
    
    /// 异步下载，支持进度流
    ///
    /// - Parameters:
    ///   - url: 流媒体 URL
    ///   - outputPath: 输出文件路径
    ///   - options: 下载选项
    /// - Returns: 下载任务，可通过 `progressStream` 获取进度
    public func downloadAsync(
        url: String,
        to outputPath: String,
        options: DownloadOptions = .default
    ) throws -> DownloadTask {
        try checkDisposed()
        
        let request = DownloadRequest(
            url: url,
            outputPath: outputPath,
            decryptionKeys: options.decryptionKeys,
            customHeaders: options.customHeaders,
            autoSelectBestQuality: options.autoSelectBestQuality
        )
        
        let encoder = JSONEncoder()
        let requestData = try encoder.encode(request)
        guard let requestJson = String(data: requestData, encoding: .utf8) else {
            throw M3U8Error.invalidParameter(name: "request")
        }
        
        let result = m3u8dl_download_async(instanceId, requestJson)
        guard result == 0 else {
            throw M3U8Error.downloadFailed(reason: "Failed to start async download")
        }
        
        return DownloadTask(downloader: self, instanceId: instanceId)
    }
    
    /// 取消当前下载
    public func cancel() {
        lock.lock()
        defer { lock.unlock() }
        
        guard !isDisposed else { return }
        m3u8dl_cancel(instanceId)
    }
    
    /// 释放下载器资源
    public func dispose() {
        lock.lock()
        defer { lock.unlock() }
        
        guard !isDisposed else { return }
        isDisposed = true
        
        progressContinuation?.finish()
        progressContinuation = nil
        
        Self.unregisterInstance(instanceId)
        m3u8dl_dispose(instanceId)
    }
    
    /// 获取进度流
    public var progressStream: AsyncStream<DownloadProgress> {
        AsyncStream { continuation in
            self.progressContinuation = continuation
            
            continuation.onTermination = { @Sendable _ in
                // 清理
            }
        }
    }
    
    /// 设置进度回调
    ///
    /// - Parameter handler: 进度回调闭包
    public func onProgress(_ handler: @escaping @Sendable (DownloadProgress) -> Void) {
        self.progressHandler = handler
    }
    
    /// 设置日志回调
    ///
    /// - Parameter handler: 日志回调闭包
    public func onLog(_ handler: @escaping @Sendable (LogLevel, String) -> Void) {
        self.logHandler = handler
    }
    
    // MARK: - Feature Query
    
    /// 检查功能是否支持
    ///
    /// - Parameter feature: 功能名称
    /// - Returns: 是否支持
    public func isFeatureSupported(_ feature: Feature) -> Bool {
        guard !isDisposed else { return false }
        return m3u8dl_is_feature_supported(instanceId, feature.rawValue) != 0
    }
    
    /// 获取可用的视频处理器列表
    ///
    /// - Returns: 处理器列表
    public func getProcessors() throws -> [Processor] {
        try checkDisposed()
        
        let resultPtr = m3u8dl_get_processors(instanceId)
        defer {
            if let ptr = resultPtr {
                m3u8dl_free_string(ptr)
            }
        }
        
        guard let ptr = resultPtr else {
            throw M3U8Error.internalError(reason: "Failed to get processors")
        }
        
        let jsonString = String(cString: ptr)
        guard let data = jsonString.data(using: .utf8) else {
            throw M3U8Error.internalError(reason: "Invalid JSON encoding")
        }
        
        return try JSONDecoder().decode([Processor].self, from: data)
    }
    
    /// 获取库版本信息
    ///
    /// - Returns: 版本信息
    public static func getVersion() -> VersionInfo? {
        let resultPtr = m3u8dl_get_version()
        defer {
            if let ptr = resultPtr {
                m3u8dl_free_string(ptr)
            }
        }
        
        guard let ptr = resultPtr else { return nil }
        
        let jsonString = String(cString: ptr)
        guard let data = jsonString.data(using: .utf8) else { return nil }
        
        return try? JSONDecoder().decode(VersionInfo.self, from: data)
    }
    
    // MARK: - Private Helpers
    
    private func checkDisposed() throws {
        lock.lock()
        defer { lock.unlock() }
        
        if isDisposed {
            throw M3U8Error.disposed
        }
    }
}

// MARK: - Callback Manager

private final class CallbackManager: @unchecked Sendable {
    // 用于管理回调的辅助类
}
